### PR TITLE
Py regex kdr

### LINKED ISSUE #2334: case_st_archive.py and env_archive.xml regexes in the context of DART
@jedwards4b @gold2718 @mvertens 

Some of the discussion in #2308 about changing regular expressions
in st_archive may be relevant to this issue.

While developing the ability to st_archive the files
created by DART, it was difficult to understand how
to use env_archive.xml.  (I have since stumbled across
www.cesm.ucar.edu/models/cesm2.0/cesm/env_archive.html,
but others might not find it, since there's no link to it
that I can find.)
There are several related issues.
1) The use of regular expression metacharacters
   makes it harder to understand.  There are no
   comments to note that they are python regexes.
   There's not enough context in the env_archive.xml
   file to understand what they need to do.
2) The inconsistency between the need for '.' separators
   in the <rest_file_extension> 'suffixes' but not
   in the <hist_file_extension> 'suffixes' because
   _archive_history_files inserts them
   (or tries to; see item 3).
3) Quite a few regular expressions that *are* in
   case_st_archive.py seem to not do quite what was
   intended.  They worked in many cases, but not as
   robustly or selectively as they should.
   These mostly involved combinations of ., \., \d and *.

I've modified scripts and xml files to move many
of the metacharacters out of env_archive.xml and
into case_st_archive.xml.
To see how this cleans up env_batch.xml, first look at
   config/config_headers.xml     (additional header comments for users)
   config/cesm/config_archive.xml
   src/drivers/mct/cime_config/config_archive.xml
   src/components/data_comps/docn/cime_config/config_archive.xml
Then look at case_st_archive.py.

I've attached a tar file of a representative CAM+POP+DART $rundir
if someone wants it for testing st_archive.  The files in it are just
placeholders, so it can be copied fast.
[CAM+POP+DART_run.gztar.zip](https://github.com/ESMCI/cime/files/1772851/CAM.POP.DART_run.gztar.zip)

### PR DESCRIPTION
**Pull request by jedwards4b**
Proposed changes to st_archive regular expressions,
which accommodate DART files, reduce metacharacters
in env_archive.xml, and make some regexes more selective
or robust.
Potential fix for #2334.
I'm not sure that e3sm changes should be in this,
but included them to be sure the regression test would work.

Test suite: scripts_regression_tests.py
Test baseline:
Test namelist changes:
Test status: bit for bit

Fixes #2334

User interface changes?: in env_archive.xml

Update gh-pages html (Y/N)?:

Code review: gold2718, jedwards4b, mfdeakin-sandia

### CODE REVIEW

File: scripts/lib/CIME/case_st_archive.py [hunk 1]
```diff
@@ -313,19 +314,20 @@ def _archive_history_files(case, archive, archive_entry,
             if ninst_string:
                 if compname.find('mpas') == 0:
                     # Not correct, but MPAS' multi-instance name format is unknown.
-                    newsuffix = compname + '.*' + suffix
+                    newsuffix =                   compname + '\d*'                   + '\.' + suffix + '\.'
```
Comment (jedwards4b) [inline scripts/lib/CIME/case_st_archive.py]: why all of the additional whitespace here?

**Comment by jedwards4b** [comment]
This is actually @kdraeder work but I opened the PR for him.

**Comment by jgfouca** [comment]
Passing the buck to @mfdeakin-sandia

File: config/cesm/config_archive.xml [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
 
   <comp_archive_spec compname="cam" compclass="atm">
-    <rest_file_extension>\.[ri]\..*</rest_file_extension>
-    <rest_file_extension>\.rh\d\.*</rest_file_extension>
-    <rest_file_extension>\.rs\.*</rest_file_extension>
-    <hist_file_extension>\.h.*.nc$</hist_file_extension>
+    <rest_file_extension>[ri]</rest_file_extension>
+    <rest_file_extension>rh\d*</rest_file_extension>
+    <rest_file_extension>rs</rest_file_extension>
+    <hist_file_extension>[eh]</hist_file_extension>
```
Comment (gold2718) [inline config/cesm/config_archive.xml]: What is the 'e' for?

File: scripts/lib/CIME/case_st_archive.py [hunk 3]
```diff
@@ -366,6 +368,12 @@ def get_histfiles_for_restarts(rundir, archive, archive_entry, restfile):
             for item in items:
                 # the following match has an option of having a './' at the beginning of
                 # the history filename
+# KDR Does this do what's intended?
+#     I think it's looking for '"...//{awordchar}{anythingelse} "'
+#     That could be done with r"\"\.*\S+\s?\""
+#     which is                 '"...{1 or more non-space}{optional space}"'
+# An example that matches the simpler pattern:
+#    "/glade/scratch/raeder/CIME_DA_vars_6/run/CIME_DA_vars_6.cam_0001.rs.2008-08-02-21600.nc"
```
Comment (gold2718) [inline scripts/lib/CIME/case_st_archive.py]: This needs to be resolved before merging.

File: scripts/lib/CIME/case_st_archive.py [hunk 4]
```diff
@@ -465,16 +473,19 @@ def _archive_restarts_date_comp(case, archive, archive_entry,
                 pfile = re.compile(pattern)
                 restfiles = [f for f in os.listdir(rundir) if pfile.search(f)]
             else:
-                pattern = r"{}\.{}\d*.*".format(casename, compname)
+#                 pattern = r"{}\.{}\d*.*".format(casename, compname)
+# KDR This finds   casename.compname[any # digits][any # non-Ret chars]
+#     Instead want casename.compname[any # digits or _].[any # non-Ret chars]
+                pattern = r"{}.{}[\d_]*\..*".format(casename, compname)
```
Comment (gold2718) [inline scripts/lib/CIME/case_st_archive.py]: This needs to be resolved before merging.

**Comment by kdraeder** [comment]
@goldy was nominated as a reviewer because the(?) resident st_archive
expert
is unavailable, @jedwards4 is the assignee, @kdraeder is the author, and we
don't know who else is interested in reviewing DART related PRs.

'e' is the file type for ESP files, some of which end up in CAM, CLM, ...
archive directories, but are not any other file type like i, r, h#, ...

I think that the 3rd item Steve flagged doesn't require much discussion.
The multi-instance file names have component parts like cam_0001,
so '_' needs to be included in the allowable characters that follow
$compname
before the next '.' .

Kevin

**Comment by gold2718** [comment]
@kdraeder, it would help if you would enter your comments on the github page where my comments are so I can be sure we are talking about the same thing. Please look there for my updates.

File: config/cesm/config_archive.xml (hunk 2 above)
Comment (gold2718) [reply config/cesm/config_archive.xml]: Do you have an example where ESP files are showing up where CAM or CLM files belong?

**Comment by kdraeder** [comment]
@goldy,
I thought that the discussion would happen under the issue,
so I reloaded https://github.com/ESMCI/cime/issues/2334
and did not see your comments, so I figured they were a direct
response to @jedwards4 original message.   I apologize for
being high maintenance again.  Now I've found the way to see
the issue comments.

**Comment by kdraeder** [comment]
@gold2718 
I removed the comments that prompted your 2nd comment,
about the restart history file regex, so that's resolved for now.  
My push put the comments into my response about not answering 
comments in the right way, which seems odd.  Is there a way that 
I can make my push comments appear as new, instead attaching 
to whatever I or someone last wrote?

**Comment by gold2718** [comment]
@kdraeder, I see the notification about your push right below the last comment which is where it is supposed to be (so to me, it looks like new top-level item).
Comments that are specifically about the PR go in the PR, not in the issue. I would phrase it as the 'Issue discussion' is to clarify the issue and discuss proposed solutions while the 'PR discussion' is for comments on the requested merge. Now I have to figure out a good place to put that on the wiki.

File: config/cesm/config_archive.xml [hunk 5]
```diff
@@ -92,40 +92,32 @@
...
-    <rest_file_extension>\.[ri]\..*</rest_file_extension>
-    <hist_file_extension>\.h\..*\.nc$</hist_file_extension>
-    <hist_file_extension>\.initial_hist\..*\.nc$</hist_file_extension>
-    <rest_history_varname>unset</rest_history_varname>
-    <rpointer>
+    <rest_file_extension>[ri]</rest_file_extension>
+    <hist_file_extension>h\d*</hist_file_extension>
+    <hist_file_extension>initial_hist</hist_file_extension>
+      <rest_history_varname>unset</rest_history_varname>
+      <rpointer>
```
Comment (gold2718) [inline config/cesm/config_archive.xml]: Are these spaces correct? I think that `<rpointer_file>` and `<rpointer_content>` should be nested under `<rpointer>` but `<rpointer>` and `<rest_history_varname>` are at the same level as `<hist_file_extension>`

File: scripts/lib/CIME/case_st_archive.py [hunk 6]
```diff
@@ -313,19 +314,20 @@ def _archive_history_files(case, archive, archive_entry,
             if ninst_string:
                 if compname.find('mpas') == 0:
                     # Not correct, but MPAS' multi-instance name format is unknown.
-                    newsuffix = compname + '.*' + suffix
+                    newsuffix =                   compname + r'\d*'                   + r'\.' + suffix + r'\.'
```
Comment (mfdeakin-sandia) [inline scripts/lib/CIME/case_st_archive.py]: Some more spacing issues

Comment (kdraeder) [reply scripts/lib/CIME/case_st_archive.py]: I did that to vertically line up the same elements in multiple lines,
to make it easier to see the differences between lines.  If this
goes against the CIME style standards, I'll remove the extra spaces.

Comment (gold2718) [reply scripts/lib/CIME/case_st_archive.py]: For the record, I find no issue with these spaces (they do increase readability IMHO).

Comment (mfdeakin-sandia) [reply scripts/lib/CIME/case_st_archive.py]: It's not my preference (personally I prefer a formatting tool to do my formatting thinking for me), but if others don't have a problem with it, then that's fine. Though the one on line 322 is missing a space.

Comment (mfdeakin-sandia) [reply scripts/lib/CIME/case_st_archive.py]: No issue, I'm fine with this PR

**Comment by kdraeder** [comment]
[eh] refers to 2 different file types.  'e' denotes files that were generated by esp (DART)
and have characteristics of a history file, but are not CAM history files.  They are closely
associated with CAM ensemble members (instances).
They are archived the same way as history files, in $DOUT_S_ROOT/atm/hist.
Example filenames are
CIME_DA_vars_11.cam_0001.e.postassim.2008-08-02-64800.nc  
CIME_DA_vars_11.cam_0001.e.preassim.2008-08-02-43200.nc   
where 'preassim' and 'postassim' refer to output from different stages
of the assimilation process.

**Comment by gold2718** [comment]
So who outputs CIME_DA_vars_11.cam_0001.e.postassim.2008-08-02-64800.nc?

**Comment by kdraeder** [comment]
The program 'filter', which is run by assimilate.csh.

**Comment by gold2718** [comment]
Then why is the entry under `<comp_archive_spec compname="cam" compclass="atm">` instead of `<comp_archive_spec compclass="esp" compname="dart">`?

**Comment by kdraeder** [comment]
Because, as described in earlier discussions about the strategy
for archiving CAM+DART files, I chose to keep DART files that 
describe an ensemble member with the CAM files that describe the
same member.  Users will want to  compare these files, and it will be 
helpful to have them in the same directory.   Also, the CESM file naming
convention and file name construction in case_st_archive.py put restrictions
on what file names can be archived by which sections of env_archive.xml.
A file name with cam_0001 will not be archived by
<comp_archive_spec compclass="esp" compname="dart">.
The file name would need to have dart_0001 in it, which would make
things more cumbersome when we do coupled assimilations which
will generate cam_0001, pop_0001, clm_0001, ... files.  If they all have
dart_0001 in their filenames, then the component part will need to be
wedged in somewhere else, making the names even longer, and no more
informative.

There are other DART files that describe the whole ensemble; those are 
archived in esp/hist.

**Comment by gold2718** [comment]
Ah, thanks.

**Pull request merged**